src/Analyzer/Passes/LogicalExpressionOptimizerPass.cpp

```diff
@@ -173,6 +173,299 @@ bool isTwoArgumentsFromDifferentSides(const FunctionNode & node_function, const
+        {
+            for (auto & flattened_argument : maybe_flattened->getArguments().getNodes())
+                new_arguments.emplace_back(std::move(flattened_argument));
+        }
+        else
+        {
+            for (auto & nested_argument : maybe_function->getArguments().getNodes())
+                new_arguments.push_back(nested_argument);
+        }
+    }
+
+    // Nothing has changed, let's no create a flattened node
+    if (!flattened_anything && new_arguments.size() == arguments.size())
+        return {};
+
+    auto flattened = std::make_shared<FunctionNode>(function_name);
+
+    flattened->getArguments().getNodes() = std::move(new_arguments);
+
+    auto function_resolver = FunctionFactory::instance().get(function_name, context);
+    flattened->resolveAsFunction(function_resolver);
+
+    return flattened;
+}
+
+struct CommonExpressionExtractionResult
+{
+    // new_node: if the new node is empty, then contain the new node, otherwise nullptr
+    // common_expressions: the extracted common expressions. The new expressions can be created
+    // as the conjunction of new_node and the nodes in common_expressions. It is guaranteed that
+    // the common expressions are deduplicated.
+    // Examples:
+    //   Input: (A & B & C) | (A & D & E)
+    //   Result: new_node = (B & C) | (D & E), common_expressions = {A}
+    //
+    //   Input: (A & B) | (A & B)
+    //   Result: new_node = nullptr, common_expressions = {A, B}
+    //
+    //   This is a special case: A & B & C is a subset of A & B, thus the conjunction of extracted
+    //   expressions is equivalent with the passed-in expression, we have to discard C. With C the
+    //   new expression would be more restrictive.
+    //   Input: (A & B) | (A & B & C)
+    //   Result: new_node = nullptr, common_expressions = {A, B}
+    QueryTreeNodePtr new_node;
+    QueryTreeNodes common_expressions;
+};
+
+std::optional<CommonExpressionExtractionResult> tryExtractCommonExpressions(const QueryTreeNodePtr & node, const ContextPtr & context)
+{
+    auto * or_node = node->as<FunctionNode>();
+    if (!or_node || or_node->getFunctionName() != "or")
+        return {}; // the optimization can only be done on or nodes
+
+    auto flattened_or_node = getFlattenedLogicalExpression(*or_node, context);
+    if (flattened_or_node)
+        or_node = flattened_or_node.get();
+
+    auto & or_argument_nodes = or_node->getArguments().getNodes();
+
+    chassert(or_argument_nodes.size() > 1);
+
+    bool first_argument = true;
+    QueryTreeNodePtrWithHashSet common_exprs_set;
+    QueryTreeNodes common_exprs;
+    QueryTreeNodePtrWithHashMap<QueryTreeNodePtr> flattened_ands;
+
+    for (auto & maybe_and_node : or_argument_nodes)
+    {
+        auto * and_node = maybe_and_node->as<FunctionNode>();
+        if (!and_node || and_node->getFunctionName() != "and")
+            return {}; // one of the nodes is not an "AND", thus there is no common expression to extract
```

Comment:
How would it work with `A | (A & B) | (A & C)`? It can be replaced with just `A`.

Of course, we can't optimize `A | (B & C) | (D & E)`

Reply:
But it can probably be done in a separate PR.

Reply:
Yes, currently single arguments won't be considered to extract. It is a reasonable expectation that it should work, so I am happy to add it in another PR, but I wouldn't like to drag this PR more than it is necessary.